### ISSUE TITLE
Intermittently happening error ("You cannot call a method on a null-valued expression.")

### ISSUE THREAD

**Issue by akunaatrium** [rmbolger/Posh-ACME#146]
Sorry for such a lousy issue report but sometimes I get this error from Posh-ACME:

```
Invoke-ACME : You cannot call a method on a null-valued expression.
At C:\ACME\Posh-ACME\Private\Update-PAOrder.ps1:55 char:29
+ ...      $response = Invoke-ACME $header ([String]::Empty) $acct -EA Stop
+                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Invoke-ACME], RuntimeExce
   ption
    + FullyQualifiedErrorId : InvokeMethodOnNull,Invoke-ACME
```

I am not sure what causes this in which circumstances, but maybe this error message gives you a hint about what could be wrong (either in Posh-ACME code or my usage of Posh-ACME).

**Comment by rmbolger** [comment]
No worries. Intermittent issues you can't reliably reproduce are super frustrating. Do you have custom scripts doing order, challenge, cert management? Or are you mostly just using the main `New-PACertificate` and `Submit-Renewal` functions?

The error message indicates the problem is in the `Update-PAOrder` function, but there are only a couple places in the code that call it and would reach that particular line. One is `Get-PAOrder` when used with the `-Refresh` parameter. The other is `Set-PAOrder` when switching between active orders.

So keep those in mind for when it happens again and perhaps we can narrow down the cause.

**Comment by akunaatrium** [comment]
I have custom scripts doing various activities like creating orders, creating HTTP challenge files if configured so (the files with the content my HTTP server must return), signaling Let's Encrypt server to execute the challenges for the orders (using `Send-ChallengeAck` for HTTP challenges and `Submit-ChallengeValidation` for DNS challenges) and then creating certificates, copying the certificates and keys to another location, etc.

Certificates are created with `New-PACertificate` function. I am not using `Submit-Renewal` function at all, although I am renewing certificates, using my own logic and the same `New-PACertificate` function).

And indeed, I am using a `Get-PAOrder` with the `-Refresh` parameter in multiple places. `Set-PAOrder` is not used anywhere.

I'll add some more logging statements to pinpoint which one of the `Get-PAOrder` calls gives the error. Let's see what happens.

**Comment by jhendricks123** [comment]
I'm extremely new to both Posh-ACME and LetsEncrypt so I apologize if this is not constructive. I am writing a Dreamhost DNS plugin, and I get this on New-PAAccount. I'm testing against LE_STAGE and have not yet tested on LE_PROD.

The error points to line 95 of New-PAAccount so I wrote the $respObj object out to see what contents were available:

`key       : @{kty=EC; crv=P-256; x=83Jph9elqyWG9EnHuAv6zw2RcMiEtn--OoRVxxz-xsI;
            y=SJyJ_m4Xy8vUvORSW_UIP1_s-63ExH7qf6kAZvs-p-8}
contact   : {}
initialIp : x.x.x.x
createdAt : 2019-06-21T19:52:07.456193105Z
status    : valid`

I think the issue in my case is specifically that the "ID" key is missing in the $respObj, so we're calling ID.ToString() and the ToString() method is obviously not present on $null.

I don't know enough about the module, ACME, or LetsEncrypt to know whether I'm using it wrong or if this is considered a bug.

**Comment by rmbolger** [comment]
Hi @jhendricks123. I'm happy to report that the specific error you're running into with `New-PAAccount` is actually fixed already in the development branch with 913be67. I just hadn't gotten around to pushing a new release with the fix in it yet. This is a [recent change](https://community.letsencrypt.org/t/acme-v2-scheduled-removal-of-id-from-account-objects/94744) with LE's server software that's currently only live on the Staging server but will be in Production soon. So I'm glad you reminded me about it. I'll try and get that out today.

If you don't want to wait, you can also install the development version using the instructions from the [readme](https://github.com/rmbolger/Posh-ACME/blob/master/README.md).

**Comment by jhendricks123** [comment]
Thanks for the rapid response! That's great news!

**Comment by danieldunderfelt** [comment]
Thank you, the update worked! I'm running this on Azure, and at first I got the same error as @jhendricks123. It was solved by importing the new version of Posh-ACME from the module gallery and republishing the runbook script.

**Comment by rmbolger** [comment]
Hey @akunaatrium. Any update on the original error you were experiencing?

**Comment by rmbolger** [comment]
I'm going to close this issue for now and hope that maybe it was just related to the ID issue in New-PAAccount. If you run into it again, feel free to re-open.

**Comment by Teazane** [comment]
Hello! Bad news, the bug seems to still be here: 
```Invoke-ACME : You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\3.7.0\Private\Update-PAOrder.ps1:55 char:29
+ ...      $response = Invoke-ACME $header ([String]::Empty) $acct -EA Stop
+                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Invoke-ACME], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull,Invoke-ACME
```

I use Posh-ACME functions in a custom script. It begins to set PA server and account (with Set-PAServer and New-PAAccount/Set-PAAccount) and then creates an order with New-PAOrder. That's the moment I get the error. I then use "Get-PAOrder | Get-PAAuthorizations" command but the error occurs just a step before at New-PAOrder. 

It worked when I tested it for the first time, and now I get this. What did I do wrong?

**Comment by rmbolger** [comment]
Hey @Teazane. It looks like you're still using version 3.7.0 and there were definitely some null related errors fixed in 3.8.0. Can you upgrade the module version and see if you can reproduce the problem consistently?  If so, feel free to open a new issue and we'll try to get to the bottom of it.
